src/Lucid/Servant.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Some helper functions for creating values in
+-- [lucid](https://hackage.haskell.org/package/lucid) DSLs that work
+-- with [servant](https://hackage.haskell.org/package/servant).
+module Lucid.Servant (
+    safeHref_,
+    safeAbsHref_,
+    safeRelHref_,
+    linkHref_,
+    linkAbsHref_,
+    linkRelHref_,
+    ) where
+
+import           Data.Proxy
+                 (Proxy)
+import           Data.Semigroup
+                 ((<>))
+import           Data.Text
+                 (Text)
+import           Lucid
+                 (Attribute)
+import           Lucid.Html5
+                 (href_)
+import           Servant.API
+                 (toUrlPiece)
+import           Servant.Utils.Links
+                 (HasLink, IsElem, Link, MkLink, safeLink')
+
+-- | 'safeLink' variant which creates lucid's 'Attribute' given base url.
+safeHref_
+    :: (IsElem endpoint api, HasLink endpoint)
+    => Text
+    -> Proxy api -> Proxy endpoint -> MkLink endpoint Attribute
+safeHref_ = safeLink' . linkHref_
+
+-- | @'safeLink' "/"@
+safeAbsHref_
+    :: (IsElem endpoint api, HasLink endpoint)
+    => Proxy api -> Proxy endpoint -> MkLink endpoint Attribute
+safeAbsHref_ = safeLink' linkAbsHref_
+
+-- | @'safeLink' ""@
+safeRelHref_
+    :: (IsElem endpoint api, HasLink endpoint)
+    => Proxy api -> Proxy endpoint -> MkLink endpoint Attribute
+safeRelHref_ = safeLink' linkRelHref_
+
+-- | Create an `href` attribute from a 'Link', with given base url.
+--
+-- "servant" ensures that any 'Link' is valid within an API.
+-- This function ensures it is possible to navigate to that endpoint from
+-- a page which shares a root with that API.
+linkHref_ :: Text -> Link -> Attribute
+linkHref_ burl = href_ . (burl <>) . toUrlPiece
```

Comment:
Rather than the warning about terminating the base url with a '/', perhaps it would be better to enforce it?

e.g. (not tested in ghci)

```haskell
slashTerminated :: Text -> Text
slashTerminated = (<> "/") . T.dropWhileEnd (== '/')
```

or

```haskell
slashTerminated :: Text -> Text
-- slashTerminated t = maybe t (<> "/") $ T.stripSuffix "/" t
slashTerminated = flip maybe (<> "/") <*> T.stripSuffix "/"
```

Reply:
yes, we can have special `<+>`, will do.

```haskell
""   <+> path = path
burl <+> path
    | T.last burl == '/' = burl <> path
    | otherwise          = burl <> "/" <> path
```